feat: Implement AlwaysRasterizeSvg

Frezyy123 opened the pull request:
https://github.com/imgproxy/imgproxy/issues/1251
* Add EnableForceRasterizeSVG option
* If enabled it skips conversion svg--->svg with unknown requested format

DarthSim commented:
Hey @Frezyy123!

Thanks for the PR!

There's a case that is not covered: if `EnableForceRasterizeSVG` is `true` but `po.SkipProcessingFormats` includes `svg`, an SVG image won't be rasterized (as expected) yet won't be sanitized and minimized.

BTW, `po.SkipProcessingFormats` check can be replaced with a one-liner: `slices.Contains(po.SkipProcessingFormats, originData.Type)`

Also, I'd name the config `AlwaysRasterizeSvg` and put it after `SanitizeSvg`.

schonert commented:
Wouldn't this be solved by simply setting the format option?

DarthSim commented:
If you know the original format, yeah, it will. Otherwise, you probably don't want to convert your JPEGs to PNGs

Frezyy123 commented:
I renamed option to AlwaysRasterizeSvg, replaced skipProcessingOptions and added sanitize everywhere
Also i wrote simple tests for some cases that cover this option

Frezyy123 commented:
btw, I see in documentation that SkipProcessing works only when the requested format is the same. But when we have unknown type it works too. 
```
Processing can only be skipped when the requested format is the same as the source format.

```
https://docs.imgproxy.net/usage/processing#skip-processing

What do you think should we divide these conditions? Or maybe we need to fix documentation?
```
if originData.Type == po.Format || po.Format == imagetype.Unknown {
```

DarthSim commented on processing_handler.go:
```diff
@@ -370,8 +371,9 @@ func handleProcessing(reqID string, rw http.ResponseWriter, r *http.Request) {
 	if originData.Type == po.Format || po.Format == imagetype.Unknown {
-		// Don't process SVG
-		if originData.Type == imagetype.SVG {
+		// Skip processing svg with unknown destination imageType
+		// if it's not forced by AlwaysRasterizeSvg option
+		if originData.Type == imagetype.SVG && !config.AlwaysRasterizeSvg {
```
It's better to check `po.SkipProcessingFormats` here instead of duplicating the whole block further

```suggestion
		shouldSkip := slices.Contains(po.SkipProcessingFormats, originData.Type)

		// Skip processing svg with unknown destination imageType
		// if it's not forced by AlwaysRasterizeSvg option
		if originData.Type == imagetype.SVG && (!config.AlwaysRasterizeSvg || shouldSkip) {
```

Frezyy123 replied:
It's not quite right condition and it fails even with old tests, but i try to rewrite it. 
However there are still duplicating with svg case
```golang
if originData.Type == imagetype.SVG && !config.AlwaysRasterizeSvg || shouldSkipProcessing {
			if originData.Type == imagetype.SVG {
```
We have two cases, svg origin data with disabled AlwaysRasterizeSvg and skipProcessing with any origin data. So our condition should work only in both cases
And then we should sanitize svg, so we have to put another check on svg and skip other formats

DarthSim replied:
This is a valid condition, you just didn't use it the way I meant. But that's ok

DarthSim commented on processing_handler_test.go:
```diff
@@ -721,6 +724,73 @@ func (s *ProcessingHandlerTestSuite) TestModifiedSinceReqCompareTooOldLastModifi
+	res := rw.Result()
+
+	require.Equal(s.T(), 200, res.StatusCode)
+	require.Equal(s.T(), "image/png", res.Header.Get("Content-Type"))
+
+	actual := s.readBody(res)
+	expected := s.readTestFile("test1.fromsvg.png")
+
+	require.True(s.T(), bytes.Equal(expected, actual))
```
It's better to just check `Content-Type` header. Otherwise, this test may fail because of changes in librsvg or libspng. Same for the rest of the tests

Frezyy123 replied:
You are right, I droped the requires and test data that I added

DarthSim commented on processing_handler_test.go:
```diff
@@ -721,6 +724,49 @@ func (s *ProcessingHandlerTestSuite) TestModifiedSinceReqCompareTooOldLastModifi
+
+func (s *ProcessingHandlerTestSuite) TestAlwaysRasterizeSvgWithEnforceAvif() {
+	config.AlwaysRasterizeSvg = true
+	config.EnforceAvif = true
```
Let's change this to WebP. The CI Docker image doesn't include libheif

DarthSim replied:
And in other tests too

Frezyy123 replied:
Changed to webp

DarthSim commented on processing_handler.go:
```diff
@@ -370,31 +371,28 @@ func handleProcessing(reqID string, rw http.ResponseWriter, r *http.Request) {
-				// it's buffer to the pool
-				originData.Close()
-
-				originData = sanitized
+		shouldSkipProcessing := slices.Contains(po.SkipProcessingFormats, originData.Type)
+
+		// Skip processing svg with unknown or the same destination imageType
+		// if it's not forced by AlwaysRasterizeSvg option
+		// Also skip processing if the format is in SkipProcessingFormats
+		if originData.Type == imagetype.SVG && !config.AlwaysRasterizeSvg || shouldSkipProcessing {
+			if originData.Type == imagetype.SVG {
+				if config.SanitizeSvg {
+					sanitized, svgErr := svg.Sanitize(originData)
+					checkErr(ctx, "svg_processing", svgErr)
+
+					// Since we'll replace origin data, it's better to close it to return
+					// it's buffer to the pool
+					originData.Close()
+
+					originData = sanitized
+				}
 			}
 
 			respondWithImage(reqID, r, rw, statusCode, originData, po, imageURL, originData)
 			return
 		}
```
If we're simplifying things here, we could've simplify 'em even more 🙂

```go
shouldSkipProcessing := (originData.Type == po.Format || po.Format == imagetype.Unknown) &&
	(slices.Contains(po.SkipProcessingFormats, originData.Type) ||
		originData.Type == imagetype.SVG && config.AlwaysRasterizeSvg)

if shouldSkipProcessing {
	if config.SanitizeSvg {
		// ... do SVG stuff
	}

	respondWithImage(reqID, r, rw, statusCode, originData, po, imageURL, originData)
	return
}
```

Frezyy123 replied:
done
Only thing that I changed is 
```
config.AlwaysRasterizeSvg --> !config.AlwaysRasterizeSvg
```

DarthSim commented:
Merged. Thanks for your contribution!
